core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java

```diff
@@ -138,13 +139,14 @@ private DeciderOutcome decide(final WorkflowDef def, final Workflow workflow, Li
 					executedTaskRefNames.remove(rt.getReferenceTaskName());
 					outcome.tasksToBeUpdated.add(task);
 					if (tasksAutoCleanup) {
-						outcome.tasksToBeDeleted = workflow.getTasks().stream()
+						List<Task> tasksToBeDeleted = workflow.getTasks().stream()
 							.filter(t -> t.getReferenceTaskName().equalsIgnoreCase(task.getReferenceTaskName())
 								&& t.getRetryCount() != 0 // Keep original
 								&& t.getRetryCount() < task.getRetryCount()) // Keep last failed
 							.collect(Collectors.toList());
-						if (!outcome.tasksToBeDeleted.isEmpty()) {
-							logger.debug("Tasks to be deleted " + outcome.tasksToBeDeleted);
+						if (!tasksToBeDeleted.isEmpty()) {
```

Comment:
Fixed a rare bug when not always all tasks picked up by auto-cleanup in one hit.